tests/integration/logs.spec.coffee

```diff
@@ -0,0 +1,186 @@
+_ = require('lodash')
+m = require('mochainon')
+Promise = require('bluebird')
+
+{ resin, givenLoggedInUser, givenMulticontainerApplication } = require('./setup')
+
+createLog = (message, device, deviceKey, options = {}) ->
+	resin.pine.post
+		resource: 'device_log'
+		body: Object.assign
+			belongs_to__device: device.id
+			device_timestamp: Date.now()
+			stderr: false
+			system: false
+		, options, { message }
+		passthrough:
+			apiKey: deviceKey
+
+createContainerLog = (message, imageInstall, deviceKey, options = {}) ->
+	resin.pine.post
+		resource: 'image_install_log'
+		body: Object.assign
+			belongs_to__image_install: imageInstall.id
+			device_timestamp: Date.now()
+			is_stderr: false
+			is_system: false
+			message: 'hi!'
+		, options, { message }
+		passthrough:
+			apiKey: deviceKey
+
+describe 'Logs', ->
+
+	givenLoggedInUser()
+
+	describe 'given an application with a api-logging device', ->
+
+		givenMulticontainerApplication()
+
+		beforeEach ->
+			Promise.all [
+				# Make sure our device looks api-log compatible
+				resin.pine.patch
+					resource: 'device'
+					body:
+						supervisor_version: '7.0.0'
+			,
+				resin.models.device.generateDeviceKey(@device.id)
+				.then (key) =>
+					@deviceKey = key
+			]
+
+		describe 'given no logs present', ->
+
+			describe '.history', ->
+
+				it 'should return no history', ->
+					resin.logs.history(@device.uuid)
+					.then (history) ->
+						m.chai.expect(history).to.deep.equal([])
+
+				it 'should emit new messages', ->
+					resin.logs.subscribe(@device.uuid)
+					.then (subscription) =>
+						messages = []
+
+						subscription.on('line', (msg) -> messages.push(msg))
+
+						createLog('Message 1', @device, @deviceKey)
+						.delay(1500)
+						.then =>
+							m.chai.expect(messages).to.have.lengthOf(1)
+							m.chai.expect(messages).to.deep.match [
+								message: 'Message 1'
+							]
+							createLog('Message 2', @device, @deviceKey)
+						.delay(1500)
+						.then ->
+							m.chai.expect(messages).to.have.lengthOf(2)
+							m.chai.expect(messages).to.deep.match [
+								message: 'Message 1'
+							,
+								message: 'Message 2'
+							]
+						.finally ->
+							subscription.unsubscribe()
+
+		describe 'given pre-existing device logs', ->
+
+			beforeEach ->
+				createLog('Test message 1', @device, @deviceKey)
+				.then =>
+					createLog 'Test message 2', @device, @deviceKey,
```

Comment:
NIT: could be `Promise.mapSeries` to have the same indentation.

Reply:
Good idea, but I tried it, and it's messier than it sounds.

I wanted to just get these as a list, and then call `Promise.runInSeries` on each of them. That doesn't exist though. You need to create a list of _functions_ that return promises, not just put these promises into a list (because as soon as you've created the promise, the operation has started), and you then need something to map over that in series and call each of those in turn.

You can do something like the below. Still, imo it's a bit too clever & less clear than the above, so I've left it for now:

```ts
Promise.mapSeries [
  () => createContainerLog('Test message 1', @newWebInstall, @deviceKey)
  () => createContainerLog('Test message 2', @newDbInstall, @deviceKey)
  () => createContainerLog('Test message 3', @newWebInstall, @deviceKey)
], (f) => f()
```